clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslServerAuthenticator.java

```diff
@@ -414,16 +542,25 @@ private boolean handleKafkaRequest(byte[] requestBytes) throws IOException, Auth
             if (apiKey != ApiKeys.API_VERSIONS && apiKey != ApiKeys.SASL_HANDSHAKE)
                 throw new IllegalSaslStateException("Unexpected Kafka request of type " + apiKey + " during SASL handshake.");
 
-            LOG.debug("Handling Kafka request {}", apiKey);
+            LOG.debug("Handling Kafka request {} during {}", apiKey, authenticationOrReauthenticationText());
 
 
             RequestContext requestContext = new RequestContext(header, connectionId, clientAddress(),
                     KafkaPrincipal.ANONYMOUS, listenerName, securityProtocol);
             RequestAndSize requestAndSize = requestContext.parseRequest(requestBuffer);
             if (apiKey == ApiKeys.API_VERSIONS)
                 handleApiVersionsRequest(requestContext, (ApiVersionsRequest) requestAndSize.request);
-            else
+            else {
                 clientMechanism = handleHandshakeRequest(requestContext, (SaslHandshakeRequest) requestAndSize.request);
+                if (previousSaslMechanism != null && !previousSaslMechanism.equals(clientMechanism)) {
```

Comment:
@rajinisivaram The `reauthenticate()` method has the `NetworkReceive` instance.  We don't parse that and extract the `SaslHandshakeRequest` that it contains until later in the flow, at this point in the code.  Is it okay to keep the changed mechanism check here, as-is?

Reply:
I just added a new `REAUTH_BAD_MECHANISM` state on the `SaslServerAuthenticator` because a change in the mechanism wasn't being recorded as a failed re-authentication in the metrics.  Now it is being recorded correctly.